### PR TITLE
Ship

### PR DESCRIPTION
**Pull request by galiptai**

### CODE REVIEW

File: src/main/java/com/codecool/spaceship/model/ship/MinerShip.java
Comment (blumlill) [inline src/main/java/com/codecool/spaceship/model/ship/MinerShip.java:1]: if we started to write custom exceptions we should keep on doing so imho

File: src/main/java/com/codecool/spaceship/model/ship/SpaceShip.java [hunk 1]
```diff
@@ -66,5 +73,7 @@ public void damageShield(int amount) {
     public double getSpeed() {
         return engine.getSpeed();
     }
+    public abstract List<ShipPart> getParts();
```
Comment (blumlill) [inline src/main/java/com/codecool/spaceship/model/ship/SpaceShip.java]: should be getPartTypes as we only tell them what kind of parts a ship has, not the actual parts

File: src/main/java/com/codecool/spaceship/model/ship/shipparts/Drill.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    );
+    private int currentLevelIndex;
+
+
+    @Override
+    public Map<Resource, Integer> getUpgradeCost() throws UpgradeNotAvailable {
+        if (getCurrentLevel() == LEVELS.size()) {
+            throw new UpgradeNotAvailable("Already at max level");
+        } else {
+            return LEVELS.get(currentLevelIndex + 1).cost();
```
Comment (blumlill) [inline src/main/java/com/codecool/spaceship/model/ship/shipparts/Drill.java:45]: if we use getcurrentlevel() in the first row, we should use it in the return too.

File: src/main/java/com/codecool/spaceship/model/ship/shipparts/ShipPart.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.codecool.spaceship.model.ship.shipparts;
+
+public enum ShipPart {
+    ENGINE,
+    SHIELD,
+    DRILL,
+    CARGO
```
Comment (blumlill) [inline src/main/java/com/codecool/spaceship/model/ship/shipparts/ShipPart.java]: should be space or storage, cargo means the actual thing they carry, not the storage space.

File: src/main/java/com/codecool/spaceship/model/ship/MinerShip.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        drill = new Drill();
+    }
+
+    public int getDrillEfficiency() {
+        return drill.getEfficiency();
+    }
+
+    @Override
+    public List<ShipPart> getParts() {
+        return List.of(ShipPart.CARGO, ShipPart.SHIELD, ShipPart.DRILL);
```
Comment (blumlill) [inline src/main/java/com/codecool/spaceship/model/ship/MinerShip.java]: should be a field imho, code is more clean that way

**Review by blumlill — APPROVED** [review]
approved but please consider my comments and suggestions to improve transparency

File: src/main/java/com/codecool/spaceship/model/ship/shipparts/Drill.java (hunk 2 above)
Comment (galiptai) [reply src/main/java/com/codecool/spaceship/model/ship/shipparts/Drill.java:45]: While they are the same thing, from a readability standpoint I think the distinction should be made. Using getCurrentLevel in the last line reads like you're getting the cost of the current level, this way it's more obvious that it's for the next one.

**Pull request merged**